File: Generation Data/Example_data_generation_for_Standard_QO.py


# Translating the provided Julia code into Python

import numpy as np

def myrandom(r, a, b):
    r = (r * 41475557) % 1
    return r, a + r * (b - a)

n = 99
dens = 0.5
dvert = 2
myseed = 1

for igen in range(5):
    r = (4 * myseed + 1) / 16384 / 16384
    myseed += 1

    Fc = np.zeros((n+1, n+1))
    Fl = np.zeros((n+1, n+1))
    F = np.zeros((n+1, n+1))

    for i in range(n):
        for j in range(i+1, n+1):
            r, num = myrandom(r, 0, 1)
            if num < dens:
                r, num = myrandom(r, 0, 10)
                Fc[i, j] = num
                Fc[j, i] = num
            else:
                r, num = myrandom(r, -10, 0)
                Fc[i, j] = num
                Fc[j, i] = num

    for i in range(n+1):
        r, num = myrandom(r, 0, dvert)
        Fl[i, i] = num

    for i in range(n):
        for j in range(i+1, n+1):
            Fl[i, j] = 0.5 * (Fl[i, i] + Fl[j, j])
            Fl[j, i] = Fl[i, j]

    for i in range(n+1):
        for j in range(i, n+1):
            F[i, j] = Fl[i, j] - Fc[i, j]
            F[j, i] = F[i, j]

    filename = f"Problem_{n+1}x{n+1}({dens})_{igen+1}.txt"
    np.savetxt(filename, F, fmt='%.15f')
    
# This code will generate 5 files with matrices. The matrices are not displayed here to save space.

File: Generation Data/Example_data_generation_for_Standard_QO.jl
%  Liuzzi et al. (2019), A new branch-and-bound algorithm for standard quadratic programming problems

function myrandom(r,a,b)
	r = mod(r*41475557.,1.)
	return (r,a+r*(b-a))
end

n      = 99
dens   = 0.9
dvert  = 2
myseed = 1

for igen = 0:5
	r = (4.*myseed + 1.)/16384./16384.
	myseed = myseed+1

	Fc = zeros(n+1,n+1)
	Fl = zeros(n+1,n+1)
	F  = zeros(n+1,n+1)

	for i = 1:n
		for j = (i+1):(n+1)
			(r,num) = myrandom(r,0,1)
			if num < dens
				(r,num) = myrandom(r,0.,10.)
				Fc[i,j] = Fc[j,i] = num
			else
				(r,num) = myrandom(r,-10.,0.)
				Fc[i,j] = Fc[j,i] = num
			end
		end
	end
	for i = 1:(n+1)
		(r,num) = myrandom(r,0,dvert)
		Fl[i,i] = num
	end
	for i = 1:n
		for j = (i+1):(n+1)
			Fl[i,j] = Fl[j,i] = 0.5*(Fl[i,i] + Fl[j,j])
		end
	end
	for i = 1:(n+1)
		for j = i:(n+1)
			F[i,j] = F[j,i] = Fl[i,j] - Fc[i,j]
		end
	end
	
	writedlm("Problem_"*string(n+1)*"x"*string(n+1)*"("*string(dens)*")_"*string(igen)*".txt",F)
end

nothing
